backend/src/sudokuboard.py

```diff
@@ -0,0 +1,137 @@
+from copy import deepcopy
+from dataclasses import dataclass
+from enum import Enum
+from functools import singledispatchmethod
+from random import shuffle, randint
+from typing import Final, List, Set
+
+
+@dataclass(frozen=True)
+class Coord:
+    x: int = 0
+    y: int = 0
+
+
+class Difficulty(Enum):
+    EASY = 32
+    MEDIUM = 27
+    HARD = 22
+    DIABOLIC = 17
+
+
+@dataclass
+class SudokuCell:
+    pos: Final[Coord]
+    value: int = 0
+    SIZE: Final[int] = 3
+
+    def __post_init__(self):
+        self.neighbors: Set[Coord] = set()
+        for n in range(self.SIZE ** 2):
+            self.neighbors.add(Coord(n, self.pos.y))
+            self.neighbors.add(Coord(self.pos.x, n))
+
+        block_coord = self._get_block_coord()
+
+        for n in range(block_coord.x, block_coord.x + self.SIZE):
+            for m in range(block_coord.y, block_coord.y + self.SIZE):
+                self.neighbors.add(Coord(n, m))
+
+        self.neighbors.remove(self.pos)
+
+    def _get_block_coord(self) -> Coord:
+        return Coord(
+            (self.pos.x // self.SIZE) * self.SIZE,
+            (self.pos.y // self.SIZE) * self.SIZE,
+        )
+
+
+class SudokuBoard:
+    def __init__(self, SIZE=3, difficulty: Difficulty = Difficulty.MEDIUM):
+        self.SIZE = SIZE
+        self.VALID_VALUES = {x + 1 for x in range(SIZE ** 2)}
+        self.cells = [
+            SudokuCell(Coord(x, y), SIZE=SIZE)
+            for x in range(SIZE ** 2)
+            for y in range(SIZE ** 2)
+        ]
+        self.difficulty = difficulty
+        self.fill_cells(0)
+
+    def reset(self):
+        for cell in self.cells:
+            cell.value = 0
+
+    def fill_cells(self, index: int) -> bool:
+        cell: SudokuCell = self.cells[index]
+
+        neighbor_values: Set[int] = {
+            self.at(neighbor).value for neighbor in cell.neighbors
+        }
+
+        options: List[int] = list(self.VALID_VALUES.difference(neighbor_values))
+
+        shuffle(options)
+
+        for option in options:
+            cell.value = option
+
+            if index == len(self.cells) - 1 or self.fill_cells(index + 1):
+                return True
+
+        cell.value = 0
+        return False
+
+    def make_puzzle(self):
+        puzzle = deepcopy(self.cells)
+        clues = self.SIZE ** 4
+        while clues >= self.difficulty.value:
+            index = randint(1, self.SIZE ** 4)  # nosec
+            value1 = puzzle[index - 1].value
+            while value1 == 0:
+                index = randint(1, self.SIZE ** 4)  # nosec
+            puzzle[index - 1].value = 0
+            self.cells = deepcopy(puzzle)
+            if not self.solve_sudoku():
+                puzzle[index - 1].value = value1
+            else:
+                clues -= 1
+
+    def solve_sudoku(self) -> bool:
+        cell = next((c for c in self.cells if c.value == 0), None)
+        if cell is None:
```

Comment:
This is a perfect example of new `walrus operator`, however *flake8* crashes when this operator is used.

Reply:
looks like pyflakes already fixed the crash